File: app/src/main/java/id/ac/ui/cs/mobileprogramming/nabilafv/rememberyourlord/ui/activity/ActListFragment.java
package id.ac.ui.cs.mobileprogramming.nabilafv.rememberyourlord.ui.activity;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ProgressBar;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProviders;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import java.util.List;

import dagger.hilt.android.AndroidEntryPoint;
import id.ac.ui.cs.mobileprogramming.nabilafv.rememberyourlord.R;
import id.ac.ui.cs.mobileprogramming.nabilafv.rememberyourlord.model.Activity;
import id.ac.ui.cs.mobileprogramming.nabilafv.rememberyourlord.utils.RecyclerViewOnItemDone;
import id.ac.ui.cs.mobileprogramming.nabilafv.rememberyourlord.utils.RecyclerViewOnItemSelected;
import id.ac.ui.cs.mobileprogramming.nabilafv.rememberyourlord.utils.State;

@AndroidEntryPoint
public class ActListFragment extends Fragment implements
        RecyclerViewOnItemDone, RecyclerViewOnItemSelected {

    private ActListViewModel actListViewModel;
    private ProgressBar progressBar;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        return  inflater.inflate(R.layout.fragment_act_list, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        actListViewModel = ViewModelProviders.of(this).get(ActListViewModel.class);

        ActListAdapter adapter = new ActListAdapter(this, this);
        actListViewModel.getAllUndoneActivity().observe(getViewLifecycleOwner(), activities -> {
            if (activities instanceof State.Success) {
                hideLoading();
                if (((State.Success) activities).data != null) {
                    adapter.setActivities((List<Activity>) ((State.Success) activities).data);
                }
            } else if (activities instanceof State.Loading) {
                showLoading();
            } else if (activities instanceof State.Failed) {
                Toast.makeText(getContext(), "Failed", Toast.LENGTH_LONG).show();
            } else if (activities instanceof State.Initialize) {
                hideLoading();
            };
        });

        RecyclerView recyclerView = view.findViewById(R.id.recyclerview_act);
        recyclerView.setLayoutManager(new LinearLayoutManager(getContext()));
        recyclerView.setHasFixedSize(true);
        recyclerView.setAdapter(adapter);

        progressBar = view.findViewById(R.id.act_list_loading);
    }

    public void hideLoading() {
        progressBar.setVisibility(View.GONE);
    }

    public void showLoading() {
        progressBar.setVisibility(View.VISIBLE);
    }

    @Override
    public void onItemSelected(Activity activity) {
        actListViewModel.selectActivity(activity);
    }

    @Override
    public void onItemDone(Activity activity) {
        actListViewModel.checkActivity(activity);
    }
}

File: app/src/main/java/id/ac/ui/cs/mobileprogramming/nabilafv/rememberyourlord/ui/activity/ActListFragment.kt
package id.ac.ui.cs.mobileprogramming.nabilafv.rememberyourlord.ui.activity

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.LinearLayoutManager
import dagger.hilt.android.AndroidEntryPoint
import id.ac.ui.cs.mobileprogramming.nabilafv.rememberyourlord.databinding.FragmentActListBinding
import id.ac.ui.cs.mobileprogramming.nabilafv.rememberyourlord.model.Activity
import id.ac.ui.cs.mobileprogramming.nabilafv.rememberyourlord.utils.RecyclerViewOnItemDone
import id.ac.ui.cs.mobileprogramming.nabilafv.rememberyourlord.utils.RecyclerViewOnItemSelected
import id.ac.ui.cs.mobileprogramming.nabilafv.rememberyourlord.utils.State


@AndroidEntryPoint
class ActListFragment : Fragment(), RecyclerViewOnItemDone, RecyclerViewOnItemSelected {
    private lateinit var actListViewModel: ActListViewModel
    private var _binding: FragmentActListBinding? = null

    private val binding get() = _binding!!

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        _binding = FragmentActListBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onDestroyView() {
        _binding = null
        super.onDestroyView()
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        actListViewModel = ViewModelProvider(requireActivity()).get(ActListViewModel::class.java)
        val adapter = ActListAdapter(this, this)
        actListViewModel.getAllUndoneActivity().observe(viewLifecycleOwner) { activities ->
            if (activities is State.Success) {
                hideLoading()
                if (activities.data != null) {
                    adapter.setActivities(activities.data as List<Activity>)
                }
            } else if (activities is State.Loading) {
                showLoading()
            } else if (activities is State.Failed) {
                Toast.makeText(context, "Failed", Toast.LENGTH_LONG).show()
            } else if (activities is State.Initialize) {
                hideLoading()
            }
        }

        binding.recyclerviewAct.layoutManager = LinearLayoutManager(context)
        binding.recyclerviewAct.setHasFixedSize(true)
        binding.recyclerviewAct.adapter = adapter
    }

    fun hideLoading() {
        binding.actListLoading.visibility = View.GONE
    }

    fun showLoading() {
        binding.actListLoading.visibility = View.VISIBLE
    }

    override fun onItemSelected(activity: Activity) {
        actListViewModel.selectActivity(activity)
    }

    override fun onItemDone(activity: Activity) {
        actListViewModel.checkActivity(activity)
    }
}